common/src/main/java/com/wynntils/features/trademarket/TradeMarketQuickSearchFeature.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright © Wynntils 2024.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features.trademarket;
+
+import com.mojang.blaze3d.platform.InputConstants;
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.core.components.Models;
+import com.wynntils.core.consumers.features.Feature;
+import com.wynntils.core.consumers.features.properties.RegisterKeyBind;
+import com.wynntils.core.keybinds.KeyBind;
+import com.wynntils.core.persisted.Persisted;
+import com.wynntils.core.persisted.config.Category;
+import com.wynntils.core.persisted.config.Config;
+import com.wynntils.core.persisted.config.ConfigCategory;
+import com.wynntils.core.text.StyledText;
+import com.wynntils.handlers.chat.event.ChatMessageReceivedEvent;
+import com.wynntils.mc.event.MenuEvent.MenuClosedEvent;
+import com.wynntils.mc.event.ScreenClosedEvent;
+import com.wynntils.mc.event.ScreenOpenedEvent;
+import com.wynntils.models.containers.containers.TradeMarketContainer;
+import com.wynntils.utils.mc.KeyboardUtils;
+import com.wynntils.utils.mc.McUtils;
+import com.wynntils.utils.wynn.ContainerUtils;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.client.gui.screens.ChatScreen;
+import net.minecraft.world.inventory.Slot;
+import net.neoforged.bus.api.EventPriority;
+import net.neoforged.bus.api.SubscribeEvent;
+import org.lwjgl.glfw.GLFW;
+
+@ConfigCategory(Category.TRADEMARKET)
+public class TradeMarketQuickSearchFeature extends Feature {
+    @Persisted
+    public final Config<Boolean> instantSearch = new Config<>(true);
+
+    @Persisted
+    public final Config<Boolean> hidePrompt = new Config<>(true);
+
+    @Persisted
+    public final Config<Boolean> autoCancel = new Config<>(true);
+
+    @RegisterKeyBind
+    private final KeyBind quickSearchKeyBind = new KeyBind(
+            "Quick Search TM",
+            GLFW.GLFW_MOUSE_BUTTON_MIDDLE,
+            InputConstants.Type.MOUSE,
+            true,
+            null,
+            this::tryQuickSearch);
+
+    private static final Pattern TYPE_TO_CHAT_PATTERN = Pattern.compile(
+            "^§5(\uE00A\uE002|\uE001) \n\uE001 Type the .* or type (\n\uE001 'cancel' to|'cancel' to \n\uE001) cancel:\n\uE001 ");
+    // Maybe there is a better solution for this than regex, but the TM is very peculiar.
+    // \\[.*?\\] Crafting stuff, \\[\\uE000-\\uF8FF\\] Gathering Tools, \\u2B21 Shiny
+    private static final Pattern CUT_PATTERN =
+            Pattern.compile("(Emerald Pouch|\\[.*?\\]|[\\uE000-\\uF8FF]+|\\u2B21)\\s*");
+
+    private static final Pattern POTION_PATTERN = Pattern.compile("^Potion of (. \\w+)");
+    private static final Pattern EOL_PATTERN = Pattern.compile("À+$");
```

Comment:
I think the use of these regexes is what bothers magicus, the fact that so many items need special handling done. 

My overall feeling is that this feature would be better if the old custom trade market searching worked, and overall, it would be great if we could wrap the trade market as a whole with a custom UI. However, I'm really afraid of doing that just because the impact of anything breaking is so significant and the code is really hard to review and manage.

Reply:
The regex seems a bit over done, but there are just some very weird edge cases with how the trade market search works. And I think I've done a pretty good job at making every item you could find on the tm searchable with quick search, tell me if I missed something or if there's a better way.
Yeah ofc this would be better with a whole custom UI, we wouldn't need use the chat and could have auto completion as someone mentioned. We’ll cross that bridge when we get there. 

I'd argue that this already provides value to the player, even if it's just QOL and not a complete overhaul.

Reply:
My problem is not with the regex existing, it's the fact that there is no explanation of the "problem" they try to solve. Does the TM not accept `[]` for pouches? Can I not search for `Potion of Health`? What do those unicode ranges stand for, what do they represent?

Don't get me wrong, I very much disagree with people writing 30 line comment-essays, most of the time, it means that the code was not properly designed, and they needed those comments to compromise on the lack of design code wise. However, in a case like this, we do need some kind of documentation on the issue itself. So, please, try to explain it, not here, but in code comments, in a reasonable way. Thank you :)